ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/tracing/flat/RewardTraceGenerator.java

```diff
@@ -0,0 +1,103 @@
+
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.results.tracing.Trace;
+import org.hyperledger.besu.ethereum.core.Block;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.core.Wei;
+import org.hyperledger.besu.ethereum.mainnet.MiningBeneficiaryCalculator;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSpec;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Stream;
+
+public class RewardTraceGenerator {
+
+  private static final String REWARD_LABEL = "reward";
+  private static final String BLOCK_LABEL = "block";
+  private static final String UNCLE_LABEL = "uncle";
+
+  /**
+   * Generates a stream of reward {@link Trace} from the passed {@link Block} data.
+   *
+   * @param protocolSchedule the {@link ProtocolSchedule} to use
+   * @param block the current {@link Block} to use
+   * @return a stream of generated reward traces {@link Trace}
+   */
+  public static Stream<Trace> generateFromBlock(
+      final ProtocolSchedule<?> protocolSchedule, final Block block) {
+
+    final List<Trace> flatTraces = new ArrayList<>();
+
+    final BlockHeader blockHeader = block.getHeader();
+    final ProtocolSpec<?> protocolSpec = protocolSchedule.getByBlockNumber(blockHeader.getNumber());
+    final MiningBeneficiaryCalculator miningBeneficiaryCalculator =
+        protocolSpec.getMiningBeneficiaryCalculator();
+
+    final AtomicReference<Wei> totalBlockReward =
+        new AtomicReference<>(protocolSpec.getBlockReward());
+
+    // add uncle reward traces
+    block
+        .getBody()
+        .getOmmers()
+        .forEach(
+            ommerBlockHeader -> {
+              final long distance = blockHeader.getNumber() - ommerBlockHeader.getNumber();
+              final Wei blockReward = protocolSpec.getBlockReward();
+              final Wei ommerReward =
+                  blockReward.subtract(blockReward.multiply(distance).divide(8));
+              final Action.Builder uncleActionBuilder =
+                  Action.builder()
+                      .author(
+                          miningBeneficiaryCalculator
+                              .calculateBeneficiary(ommerBlockHeader)
+                              .toHexString())
+                      .rewardType(UNCLE_LABEL)
+                      .value(ommerReward.toShortHexString());
+              flatTraces.add(
+                  RewardTrace.builder()
+                      .actionBuilder(uncleActionBuilder)
+                      .blockHash(block.getHash().toHexString())
+                      .blockNumber(blockHeader.getNumber())
+                      .type(REWARD_LABEL)
+                      .build());
+
+              // add uncle inclusion reward
+              totalBlockReward.set(totalBlockReward.get().add(blockReward.divide(32)));
```

Comment:
Same, is 32 a magic number? Should that logic be tested elsewhere? It may be impacted by the chain spec (ETH vs ETC have different block rewards)

Reply:
The same I just made a modification that uses the blockProcessor and I added tests to verify that the RewardTracegenerator is using the processor to calculate the rewards